### PR TITLE
Use a SmallVec for the animated tile list

### PR DESCRIPTION
**Pull request by michicc**
One resizeable list implementation is enough.

### CODE REVIEW

File: src/core/smallvec_type.hpp [hunk 1]
```diff
@@ -241,6 +258,19 @@ class SmallVector {
 		if (to_move > 0) MemMoveT(this->data + pos, this->data + pos + count, to_move);
 	}
 
+	/**
+	 * Extract an item from this vector, moving the following items to fill the gap.
```
Comment (frosch123) [inline src/core/smallvec_type.hpp]: This method seems to differ from Erase by not changing the order of the items.
Is this the intention, is it needed?
If yes, the docs should point that out. If not, use Erase :)

Comment (frosch123) [reply src/core/smallvec_type.hpp]: There is also ErasePreservingOrder.
Maybe an overload of that method?

File: src/animated_tile.cpp [hunk 2]
```diff
@@ -52,27 +42,17 @@ void AddAnimatedTile(TileIndex tile)
...
-	/* Table not large enough, so make it larger */
-	if (_animated_tile_count == _animated_tile_allocated) {
-		_animated_tile_allocated *= 2;
-		_animated_tile_list = ReallocT<TileIndex>(_animated_tile_list, _animated_tile_allocated);
-	}
-
-	_animated_tile_list[_animated_tile_count] = tile;
-	_animated_tile_count++;
+	if (_animated_tiles.Contains(tile)) return;
+	*_animated_tiles.Append() = tile;
```
Comment (frosch123) [inline src/animated_tile.cpp]: These two lines can use Include() instead.

File: src/core/smallvec_type.hpp (hunk 1 above)
Comment (michicc) [reply src/core/smallvec_type.hpp]: ErasePreservingOrder must be new (... well, newer than this originally stone age old commit :)

Comment (frosch123) [reply src/core/smallvec_type.hpp]: 5 years :)

**Comment by michicc** [comment]
Make use of new SmallVec functions to reduce patch.

File: src/core/smallvec_type.hpp [hunk 3]
```diff
@@ -158,6 +158,23 @@ class SmallVector {
 		}
 	}
 
+	/**
+	 * Insert a new item at a specific position into the vector, moving all following items.
+	 * @param item Position at which the new item should be inserted
+	 * @return pointer to the new item
+	 */
+	inline T *Insert(T *item)
```
Comment (TrueBrain) [inline src/core/smallvec_type.hpp:166]: You introduce this function, but you don't use it; was that intended?

File: src/animated_tile.cpp [hunk 4]
```diff
@@ -51,28 +42,16 @@ void DeleteAnimatedTile(TileIndex tile)
...
 
 /**
  * Animate all tiles in the animated tile list, i.e.\ call AnimateTile on them.
  */
 void AnimateAnimatedTiles()
 {
-	const TileIndex *ti = _animated_tile_list;
-	while (ti < _animated_tile_list + _animated_tile_count) {
+	const TileIndex *ti = _animated_tiles.Begin();
+	while (ti < _animated_tiles.End()) {
```
Comment (TrueBrain) [inline src/animated_tile.cpp:54]: Shouldn't this be `!=`? `<` for pointer arithmetic feels a bit weird?

File: src/core/smallvec_type.hpp [hunk 5]
```diff
@@ -232,13 +249,24 @@ class SmallVector {
 	 * @param count Number of consecutive items to remove.
 	 */
 	void ErasePreservingOrder(uint pos, uint count = 1)
+	{
+		ErasePreservingOrder(this->data + pos, count);
```
Comment (TrueBrain) [inline src/core/smallvec_type.hpp:253]: `this->Begin() + pos`?

File: src/saveload/afterload.cpp [hunk 6]
```diff
@@ -2157,22 +2157,21 @@ bool AfterLoadGame()
...
-		for (uint i = 0; i < _animated_tile_count; /* Nothing */) {
+		for (TileIndex *tile = _animated_tiles.Begin(); tile != _animated_tiles.End(); /* Nothing */) {
 			/* Remove if tile is not animated */
-			bool remove = _tile_type_procs[GetTileType(_animated_tile_list[i])]->animate_tile_proc == NULL;
+			bool remove = _tile_type_procs[GetTileType(*tile)]->animate_tile_proc == NULL;
 
 			/* and remove if duplicate */
-			for (uint j = 0; !remove && j < i; j++) {
-				remove = _animated_tile_list[i] == _animated_tile_list[j];
+			for (TileIndex *j = _animated_tiles.Begin(); !remove && j < tile; j++) {
```
Comment (TrueBrain) [inline src/saveload/afterload.cpp:2167]: Here too, possibly replace `<` with `!=`? Don't know if the SmartVec guarantees memory to be in sequence :) (seems so, realloc, but it reads a bit weird) :D

**Review by TrueBrain — COMMENTED** [review]
The CI fail made me curious what your patch changes in the randomizer; my head now spins, but it did leave me with a few questions :) Possibly a bit out of scope for your patch; I was just curious :D

File: src/core/smallvec_type.hpp (hunk 3 above)
Comment (michicc) [reply src/core/smallvec_type.hpp:166]: Yes and no :) This is some old backlog cleanup, and `Insert` is used in some other commits.

File: src/animated_tile.cpp (hunk 4 above)
Comment (michicc) [reply src/animated_tile.cpp:54]: The while loop can shorten the tile list, so the `<` was probably some safeguard. I decided to leave that alone.

File: src/core/smallvec_type.hpp (hunk 5 above)
Comment (michicc) [reply src/core/smallvec_type.hpp:253]: Question of taste I guess.

**Comment by michicc** [comment]
I left the `Insert` in, but it can be removed if preferred.

**Comment by TrueBrain** [comment]
I am good with it; @frosch123 did you had any other points?

**Review by frosch123 — APPROVED** [review]

**Pull request merged**